crypto/crypto/src/macros/src/lib.rs

```diff
@@ -36,3 +111,327 @@ pub fn silent_debug(source: TokenStream) -> TokenStream {
+        Data::Enum(ref variants) => impl_enum_validkey(name, variants),
+        Data::Struct(_) | Data::Union(_) => panic!("#[derive(ValidKey)] is only defined for enums"),
+    }
+}
+
+fn impl_enum_tryfrom(name: &Ident, variants: &DataEnum) -> QuoteTokenStream {
+    // the TryFrom dispatch
+    let mut try_iter = variants.variants.iter();
+    let first_variant = try_iter
+        .next()
+        .expect("#[derive(ValidKey] requires a non-empty enum.");
+    let first_variant_ident = &first_variant.ident;
+    let first_variant_arg = &first_variant
+        .fields
+        .iter()
+        .next()
+        .expect("Unrecognized enum for key types")
+        .ty;
+
+    let mut try_chain = quote! {
+        #first_variant_arg::try_from(bytes).and_then(|key| Ok(#name::#first_variant_ident(key)))
+    };
+    for variant in try_iter {
+        let variant_ident = &variant.ident;
+        let variant_arg = &variant
+            .fields
+            .iter()
+            .next()
+            .expect("Unrecognized enum for key types")
+            .ty;
+        try_chain.extend(quote!{
+            .or_else(|_err| #variant_arg::try_from(bytes).and_then(|key| Ok(#name::#variant_ident(key))))
+        })
+    }
+
+    quote! {
+        impl core::convert::TryFrom<&[u8]> for #name {
+            type Error = crypto::CryptoMaterialError;
+            fn try_from(bytes: &[u8]) -> std::result::Result<#name, Self::Error> {
+                #try_chain
+            }
+        }
+    }
+}
+
+fn match_enum_to_bytes(name: &Ident, variants: &DataEnum) -> QuoteTokenStream {
+    // the ValidKey dispatch proper
+    let mut match_arms = quote! {};
+    for variant in variants.variants.iter() {
+        let variant_ident = &variant.ident;
+
+        match_arms.extend(quote! {
+            #name::#variant_ident(key) => key.to_bytes().to_vec(),
+        });
+    }
+    match_arms
+}
+
+fn impl_enum_validkey(name: &Ident, variants: &DataEnum) -> TokenStream {
+    let mut try_from = impl_enum_tryfrom(name, variants);
+
+    let to_bytes_arms = match_enum_to_bytes(name, variants);
+
+    try_from.extend(quote! {
+
+        impl ::crypto::ValidKey for #name {
+            fn to_bytes(&self) -> Vec<u8> {
+                match self {
+                    #to_bytes_arms
+                }
+            }
+        }
+    });
+    try_from.into()
+}
+
+#[proc_macro_derive(PublicKey, attributes(PrivateKeyType))]
+pub fn derive_enum_publickey(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let name = &ast.ident;
+    let private_key_type = get_type_from_attrs(&ast.attrs, "PrivateKeyType");
+    match ast.data {
+        Data::Enum(ref variants) => impl_enum_publickey(name, private_key_type, variants),
+        Data::Struct(_) | Data::Union(_) => {
+            panic!("#[derive(PublicKey)] is only defined for enums")
+        }
+    }
+}
+
+fn get_type_from_attrs(attrs: &[syn::Attribute], attr_name: &str) -> String {
```

Comment:
Return type should be `syn::Result<String>` since we would ideally produce an error if the user writes the attribute wrong. Better, it should be `syn::Result<syn::LitStr>` so that when you later parse the string, errors in that second parse end up pointing to the right input token.